src/ns/core/notifier.py

```diff
@@ -115,60 +116,69 @@ async def send_notification(self, *, notification: event_schemas.Notification):
+    ):
+        """Builds an email message subtype (HTML or plaintext) from a template and
+        a dictionary of values.
+        """
+        template_str = (
+            self._config.plaintext_email_template
+            if template_type == EmailTemplateType.PLAINTEXT
+            else self._config.html_email_template
+        )
 
-        payload_as_dict = notification.model_dump()
+        # Load the string as a python string template
+        template = Template(template_str)
 
-        # create plaintext html with template
-        plaintext_template = Template(self._config.plaintext_email_template)
+        # Try to substitute the values into the template
         try:
-            plaintext_email = plaintext_template.substitute(payload_as_dict)
+            email_subtype = template.substitute(values_dict)
         except KeyError as err:
             template_var_error = self.VariableNotSuppliedError(variable=err.args[0])
             log.critical(template_var_error, extra={"variable": err.args[0]})
             raise template_var_error from err
         except ValueError as err:
             template_format_error = self.BadTemplateFormat(
-                template_type=EmailTemplateType.PLAINTEXT, problem=err.args[0]
+                template_type=template_type, problem=err.args[0]
             )
             log.critical(
                 template_format_error,
-                extra={
-                    "template_type": EmailTemplateType.PLAINTEXT,
-                    "problem": err.args[0],
-                },
+                extra={"template_type": template_type, "problem": err.args[0]},
             )
             raise template_format_error from err
+        return email_subtype
 
+    def _construct_email(
+        self, *, notification: event_schemas.Notification
+    ) -> EmailMessage:
+        """Constructs an EmailMessage object from the contents of an email notification event"""
+        message = EmailMessage()
+        message["To"] = notification.recipient_email
+        message["Cc"] = notification.email_cc
+        message["Bcc"] = notification.email_bcc
+        message["Subject"] = notification.subject
+        message["From"] = self._config.from_address
+
+        payload_as_dict = {**notification.model_dump()}
+
+        # create plaintext html with template
+        plaintext_email = self._build_email_subtype(
+            template_type=EmailTemplateType.PLAINTEXT, values_dict=payload_as_dict
+        )
         message.set_content(plaintext_email)
 
-        # create html version of email, replacing variables of $var format
-        html_template = Template(self._config.html_email_template)
+        # Escape values exposed to the email in case they've been maliciously crafted
+        for k, v in payload_as_dict.items():
+            if isinstance(v, list):
+                payload_as_dict[k] = ", ".join([html.escape(_) for _ in v])
```

Comment:
By convention _ would be used as variable name, if it should be ignored, e.g. unpacking a tuple and only using the first two of three values and ignoring the last one.
A proper variable name should be used here instead.

Reply:
Also, not sure if escaping should be on the caller of the code that actually does the substitution.
`template_type` is passed along, so this could be moved to `_build_email_subtype` and run conditionally based on that

Reply:
Good points. I've moved the escape chunk into that method and renamed both the variables you highlighted.